### ISSUE TITLE
Border instance should not change depending on parent

### ISSUE THREAD

**Issue by Chrriis** [JFormDesigner/FlatLaf#36]
When a button is placed in a toolbar, the appearance of its border changes. I found that the border is replaced when parenting to the toolbar, while it should be the same instance with a different painting logic. Otherwise, compounding the default border with another border does not work.

Here is a simple test case to illustrate:
```Java
    public static void main(String[] args) throws Exception {
        UIManager.setLookAndFeel(new FlatLightLaf());
        JFrame frame = new JFrame();
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        JPanel panel = new JPanel();
        JToolBar toolBar = new JToolBar();
        JButton button1 = new JButton(new FlatFileViewFloppyDriveIcon());
        toolBar.add(button1);
        JButton button2 = new JButton(new FlatFileViewFloppyDriveIcon());
        button2.setBorder(BorderFactory.createCompoundBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10), button2.getBorder()));
        toolBar.add(button2);
        panel.add(toolBar);
        frame.getContentPane().add(panel);
        frame.setSize(400, 300);
        frame.setVisible(true);
    }
```

Here is a screenshot of the result:
![BorderCompounding](https://user-images.githubusercontent.com/1174202/72070278-703c7300-32e9-11ea-9e9b-a1e0a40e4539.png)

Notice that the second button has a border that is painted, which it should not have.

**Comment by j-dimension** [comment]
@j-dimension putting this on my watch list also. Seen the same behavior.

**Comment by DevCharly** [comment]
This is default behavior of JToolBar (in all LAFs).

The border of added buttons (if instance of `AbstractButton`) is replaced with a special border for toolbars, if the button border is an instance of `UIResource`. This is done in `BasicToolBarUI`.

In your example the border of the first button is replaced because `FlatButtonBorder` implements `UIResouce`. The border of the second button is not replaced.

Using compound borders to add extra space is not necessary. Better use:
~~~java
button2.setMargin( new Insets( 10, 10, 10, 10 ) );
~~~

To change margins of all toolbar buttons use:
~~~java
UIManager.put( "ToolBar.buttonMargins", new Insets( 5, 5, 5, 5 ) );
~~~

**Comment by Chrriis** [comment]
I disagree. On Windows L&F, a button in a toolbar and a regular button paint their border differently, but the instance of the border is the same. It is a `com.sun.java.swing.plaf.windows.XPStyle$XPEmptyBorder`. Check its implementation, it handles various types of parents.

Adding spaces with a compound border is a basic example, there are all sorts of other compounding cases that can be done. But as for changing the margin, adding an empty border around allows to add a fixed number without having to do math computations on the existing margin of the component under the current L&F.

**Comment by Chrriis** [comment]
Coming back on this issue as it is a pain for me 😉 

I think there are many Java applications that were made for Windows only, and Windows L&F allows compounding borders. Developers of such applications do not realize that other L&F do not have this feature. Having to review all border logic is a pain (especially identifying all the places where borders are tweaked).

It would make adoption of FlatLaf much easier and smooth if there was a single border class like on Windows L&F.

**Comment by Chrriis** [comment]
In case someone needs it, here is some code that replaces the border instances of the buttons with a stable version that behaves differently depending on the parent. It contains a main method to test too.
It is a bit dirty, but I did not find a better way with FlatLaf current API. Ideas welcome...
```Java
	private static void installStableButtonBorders() {
		UIDefaults uiDefaults = UIManager.getDefaults();
		Border toggleButtonBorder = uiDefaults.getBorder("ToggleButton.border");
		Border buttonBorder = uiDefaults.getBorder("Button.border");
		class ExtendedFlatToolBarUI extends FlatToolBarUI {
			@Override
			public Border getNonRolloverBorder(AbstractButton b) {
				return super.getNonRolloverBorder(b);
			}
			@Override
			public Border getRolloverBorder(AbstractButton b) {
				return super.getRolloverBorder(b);
			}
		}
		Border stableButtonBorder = new Border() {
			private ExtendedFlatToolBarUI extendedFlatToolBarUI = new ExtendedFlatToolBarUI();
			@Override
			public void paintBorder(Component c, Graphics g, int x, int y, int width, int height) {
				getRealBorder(c).paintBorder(c, g, x, y, width, height);
			}
			@Override
			public Insets getBorderInsets(Component c) {
				return getRealBorder(c).getBorderInsets(c);
			}
			@Override
			public boolean isBorderOpaque() {
				return false;
			}
			private Border getRealBorder(Component c) {
				if(!(c instanceof AbstractButton)) {
					return buttonBorder;
				}
				if(c.getParent() instanceof JToolBar && ((JToolBar)c.getParent()).getUI() instanceof FlatToolBarUI) {
					boolean isRolloverBorders = ((FlatToolBarUI)((JToolBar)c.getParent()).getUI()).isRolloverBorders();
					if(isRolloverBorders) {
						return extendedFlatToolBarUI.getRolloverBorder((AbstractButton)c);
					}
					return extendedFlatToolBarUI.getNonRolloverBorder((AbstractButton)c);
				}
				if(c instanceof JToggleButton) {
					return toggleButtonBorder;
				}
				return buttonBorder;
			}
		};
		uiDefaults.put("ToggleButton.border", stableButtonBorder);
		uiDefaults.put("Button.border", stableButtonBorder);
	}
	public static void main(String[] args) throws Exception {
		UIManager.setLookAndFeel(new FlatLightLaf());
		// Make the borders of JButton and JToggleButton stable
		installStableButtonBorders();
		JFrame frame = new JFrame();
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		JPanel panel = new JPanel(new FlowLayout());
		JToolBar toolBar = new JToolBar();
		JButton button1 = new JButton(new FlatFileViewFloppyDriveIcon());
		toolBar.add(button1);
		JButton button2 = new JButton(new FlatFileViewFloppyDriveIcon());
		button2.setBorder(BorderFactory.createCompoundBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10), button2.getBorder()));
		toolBar.add(button2);
		panel.add(toolBar);
		panel.add(new JButton("Button3"));
		panel.add(new JButton("Button4"));
		panel.add(new JToggleButton("Button5"));
		frame.getContentPane().add(panel);
		frame.setSize(400, 300);
		frame.setVisible(true);
	}
```

One thing I am not sure about is the fact that `FlatButtonUI` has some painting logic dependent on the class of the button border (which is replaced by above code).

**Comment by DevCharly** [comment]
@Chrriis I've changed the implementation. It no longer uses rollover borders.

Please give it a try. See master branch.

**Comment by Chrriis** [comment]
@DevCharly It works great and it allows me to remove all my workaround junk! Thanks a lot! 🙂

**Comment by DevCharly** [comment]
released in 0.27
